sdk/core/Azure.Core/src/Pipeline/HttpPipeline.cs

```diff
@@ -66,11 +96,23 @@ public HttpMessage CreateMessage()
         /// <param name="message">The <see cref="HttpMessage"/> to send.</param>
         /// <param name="cancellationToken">The <see cref="CancellationToken"/> to use.</param>
         /// <returns>The <see cref="ValueTask"/> representing the asynchronous operation.</returns>
-        public ValueTask SendAsync(HttpMessage message, CancellationToken cancellationToken)
+        public async ValueTask SendAsync(HttpMessage message, CancellationToken cancellationToken)
         {
             message.CancellationToken = cancellationToken;
             AddHttpMessageProperties(message);
-            return _pipeline.Span[0].ProcessAsync(message, _pipeline.Slice(1));
+
+            if (message.Policies == null)
+            {
+                await _pipeline.Span[0].ProcessAsync(message, _pipeline.Slice(1)).ConfigureAwait(false);
+            }
+            else
+            {
+                var length = _pipeline.Length + message.Policies.Count;
+                var policies = ArrayPool<HttpPipelinePolicy>.Shared.Rent(length);
```

Comment:
The rent/return needs to be wrapped in try-finally, i.e. return needs to happen in finally.